A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,207 @@
+
+### Subchannel Wrapper
+
+The subchannel wrappers created by the `outlier_detection` LB policy will hold a reference to its map entry in the LB policy, if that map entry exists.
+
+The subchannel wrapper will track the latest state update from the underlying subchannel. By default, it will simply pass those updates along. The wrapper will have the following methods:
+
+ - `eject()`: The wrapper will report a state update with the `TRANSIENT_FAILURE` state, and will stop passing along updates from the underlying subchannel.
+ - `uneject()`: The wrapper will report a state update with the latest update from the underlying subchannel, and resume passing along updates from the underlying subchannel.
+
+### Success Rate Algorithm
+
+ 1. If the number of addresses with request volume of at least `success_rate_request_volume` is less than `success_rate_minimum_hosts`, stop.
+ 2. Calculate the mean and standard deviation of the fractions of successful requests among addresses with total request volume of at least `success_rate_request_volume`.
+ 3. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `success_rate_request_volume`, continue to the next address.
+    3. If the address's success rate is less than `(mean - stdev * (success_rate_stdev_factor / 1000))`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_success_rate`, eject that address.
+
+### Failure Percentage Algorithm
+
+ 1. If the number of addresses is less than `failure_percentage_minimum_hosts`, stop.
+ 2. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `failure_percentage_request_volume`, continue to the next address.
+    3. If the address's failure percentage is greater than `failure_percentage_threshold`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_failure_percentage`, eject that address.
+
+### xDS Integration
+
+If the `Cluster` message for an underlying cluster has the `outlier_detection` field set, the `xds_cluster_resolver` will generate an `outlier_detection` LB policy config from it for each priority as the top-level LB policy for that priority. In that case, the child policy config will be the config that would otherwise be at the top level for the priority. When `outlier_detection` is configured, the LB policy tree will look like this when using round robin, and the `outlier_detection` LB policy will be inserted similarly when using ring hash:
```

Comment:
If we insert the `outlier_detection` LB policy in the tree only when outlier detection is configured in xDS, then a config change that enables or disables outlier detection will cause churn in the LB policy tree.  This is more of an issue in Java and Go than it is in C-core and Node, because the former two implementations do not have a subchannel pool and will therefore have to reconnect to all of the endpoints when this reconfiguration occurs.  But even in C-core and Node, there may be cases where child policies will lose some state when this reconfiguration occurs, which may not be optimal.

We could avoid this by saying that we always insert the outlier detection LB policy, even when outlier detection is not configured, in which case the policy is a no-op.  But this would come at the cost of having that policy in the tree even when it's not needed.

@ejona86 and @dfawley, thoughts?

Reply:
I went ahead and made this change.

Reply:
Okay.  I'd like feedback from Eric and Doug before finalizing, though.  The counter-argument to what I said above is that config changes to enable or disable outlier detection are probably quite rare, and it may not be worth imposing the extra overhead of the outlier detection policy when it's not being used.

Reply:
SGTM; the overhead of the extra policy is pretty low, and if the reconnections are triggered, they will all happen at once, which seems pretty invasive.

Reply:
I'd prefer the tree to be as static as possible. That means config changes would be less invasive (less likely to trigger bug) and fewer code paths to consider for an RPC. The noop cost should be low, and could be lower than using it conditionally (because of branch prediction and JITs noticing a call site only calls one particular implementation).

As much as I like "not paying for things you aren't using," xDS _is_ using outlier detection, so I'd consider avoiding it premature optimization.